cmd/webhook/main.go

```diff
@@ -17,62 +17,65 @@ import (
 	kms "cloud.google.com/go/kms/apiv1"
 	secretmanager "cloud.google.com/go/secretmanager/apiv1"
 	"cloud.google.com/go/secretmanager/apiv1/secretmanagerpb"
-	"github.com/bradleyfalzon/ghinstallation/v2"
 	"github.com/chainguard-dev/clog"
 	metrics "github.com/chainguard-dev/terraform-infra-common/pkg/httpmetrics"
-	"github.com/kelseyhightower/envconfig"
-	"github.com/octo-sts/app/pkg/gcpkms"
+	envConfig "github.com/octo-sts/app/pkg/envconfig"
+	"github.com/octo-sts/app/pkg/ghtransport"
 	"github.com/octo-sts/app/pkg/webhook"
 )
 
-type envConfig struct {
-	Port          int    `envconfig:"PORT" required:"true" default:"8080"`
-	KMSKey        string `envconfig:"KMS_KEY" required:"true"`
-	AppID         int64  `envconfig:"GITHUB_APP_ID" required:"true"`
-	WebhookSecret string `envconfig:"GITHUB_WEBHOOK_SECRET" required:"true"`
-}
-
 func main() {
 	ctx, cancel := signal.NotifyContext(context.Background(), os.Interrupt)
 	defer cancel()
 	ctx = clog.WithLogger(ctx, clog.New(slog.Default().Handler()))
 
-	go metrics.ServeMetrics()
-
-	// Setup tracing.
-	defer metrics.SetupTracer(ctx)()
-
-	var env envConfig
-	if err := envconfig.Process("", &env); err != nil {
+	baseCfg, err := envConfig.BaseConfig()
+	if err != nil {
 		log.Panicf("failed to process env var: %s", err)
 	}
-
-	kms, err := kms.NewKeyManagementClient(ctx)
+	webhookConfig, err := envConfig.WebhookConfig()
 	if err != nil {
-		log.Panicf("could not create kms client: %v", err)
+		log.Panicf("failed to process env var: %s", err)
 	}
-	signer, err := gcpkms.New(ctx, kms, env.KMSKey)
-	if err != nil {
-		log.Panicf("error creating signer: %v", err)
+
+	if baseCfg.Metrics {
+		go metrics.ServeMetrics()
+
+		// Setup tracing.
+		defer metrics.SetupTracer(ctx)()
 	}
-	atr, err := ghinstallation.NewAppsTransportWithOptions(http.DefaultTransport, env.AppID, ghinstallation.WithSigner(signer))
+
+	var client *kms.KeyManagementClient
+
+	if baseCfg.KMSKey != "" {
+		client, err = kms.NewKeyManagementClient(ctx)
+		if err != nil {
+			log.Panicf("could not create kms client: %v", err)
+		}
+	}
+
+	atr, err := ghtransport.New(ctx, baseCfg, client)
 	if err != nil {
 		log.Panicf("error creating GitHub App transport: %v", err)
 	}
 
 	webhookSecrets := [][]byte{}
-	secretmanager, err := secretmanager.NewClient(ctx)
-	if err != nil {
-		log.Panicf("could not create secret manager client: %v", err)
-	}
-	for _, name := range strings.Split(env.WebhookSecret, ",") {
-		resp, err := secretmanager.AccessSecretVersion(ctx, &secretmanagerpb.AccessSecretVersionRequest{
-			Name: name,
-		})
+	if baseCfg.KMSKey != "" {
```

Comment:
the web hook secret in the original code uses the secretmanager which is google KMS specific? If you're not using the KMSKey with google, then the web hook secret can be inputted as an env var. Does that make sense or am I off?

Reply:
It's a little odd because they are technically independent of each other - KMS is used for outbound auth, whereas the webhook secret is use for inbound validation.

I see what you're saying though about it's a good indication that you want to enable Cloud provider support. I suspect we might want to add a separate envvar (or prefix the kms/secret names with a prefix scheme), but for now I'm okay with moving forward with this as-is.

Could you add a comment here with a short description for why we're doing this? 🙏

Reply:
> Could you add a comment here with a short description for why we're doing this? 🙏

comment added